README.md
# influxdb-csv-cleaner

Tool to clean CSV content exported from [InfluxDB](https://www.influxdata.com/).

## What it does

 - Remove repeated header lines, appearing in the middle of file.
 - Remove first column, which is just table (measurement) name.
 - Convert timestamp in "time" column to a readable format, like "2018-01-30 00:00:00".

## Build

1. For BeagleBone

    - Install _gcc-6-arm-linux-gnueabihf_, _binutils-arm-linux-gnueabihf_ packages.
    - Install _armv7-unknown-linux-gnueabihf_ target for Rust:

    ```sh
    rustup target install armv7-unknown-linux-gnueabihf
    ```

    - If you are cross-compiling from PC, need to add this content to _.cargo/config_ file:

    ```toml
    [target.armv7-unknown-linux-gnueabihf]
    linker = "arm-linux-gnueabihf-gcc"
    ```

    - Build with command:

    ```sh
    cargo build --target=armv7-unknown-linux-gnueabihf --release
    ```

    - Strip the compiled file to reduce file size:

    ```sh
    /usr/arm-linux-gnueabihf/bin/strip target/armv7-unknown-linux-gnueabihf/release/influxdb-csv-cleaner
    ```

2. For PC

Too simple to tell.

## Example

You export data from InfluxDB with this command:

```sh
influx -database myfarm -precision s -format csv -execute "SELECT temperature FROM condition LIMIT 100"
```

And save to a file _sample.csv_:

```csv
    name,time,temperature
    condition,1489544029,29.1
    condition,1489544039,29.2
```

Now, you want to remove the first column:

```sh
influxdb-csv-cleaner sample.csv -t Asia/Ho_Chi_Minh -o clean.csv
```

The ouput will be:

```csv
    time,temperature
    2017-03-15 09:13:49,29.1
    2017-03-15 09:13:59,29.2
```

The timestamp column is always convert to readable format. It also help avoid confusing when you use this CSV with graphing tool.

Note: The header line can apear many times in the InfluxDB export file, because `influx` client then makes chunked queries to handle big data. But `influxdb-csv-cleaner` tool will skip all of them, except the top line.

You can also use the tool in pipeline to clean on the _stdin_ stream:

```sh
influx -database myfarm -precision s -format csv -execute "SELECT temperature FROM condition LIMIT 100" | influxdb-csv-cleaner - -t Asia/Ho_Chi_Minh
```

Please run

```sh
influxdb-csv-cleaner -h
```

to see more other usages.


CHANGELOG.md
# Changelog

## 0.2.3
* Learn new Rust error handling technique.

## 0.2.2
* Bring back timezone conversion feature.

## 0.2.1
* Add more info to help build debian package.
* (Internal) Drop use of "human-panic". It is not the thing I need.

## 0.2.0
* Remove timezone conversion feature, because we now can ask InfluxDB to export data in desired timezone.

## 0.1.1
* Let filter other rows, keep only rows which "time" part of the day match a specified value, like "07:00:00".

## 0.1.0
* First version


src/main.rs
//! Tool to clean CSV content exported from InfluxDB.
//!
//! Sample input:
//!
//!    name,time,temperature,crop_id
//!    condition,1489544029,29.1,5
//!    condition,1489544039,29.2,5
//!
//! Output with timezone Asia/Ho_Chi_Minh:
//!
//!    time,temperature,crop_id
//!    2017-03-15 09:13:49 +07,29.1,5
//!    2017-03-15 09:13:59 +07,29.2,5
//!

use std::io;
use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;
use std::io::BufWriter;
use std::io::Write;
use std::path::Path;

use clap::{Arg, App};
use clap::{crate_version, crate_authors};
use chrono::{TimeZone, NaiveTime};
use chrono_tz::Tz;
use color_eyre::eyre::Result;
use eyre::{WrapErr, eyre};


fn concat_columns(last: String, current: &str) -> String {
	last + "," + current
}


fn process_line(line: String, on_first_line: bool, dest_timezone: Tz, time_point: Option<NaiveTime>, quiet: bool) -> Option<String> {
	let mut column_iter = line.split(',');
	// First column is measurement name. Skip it
	column_iter.next();
	// The second column becomes the first, and should contain timestamp in UTC
	let date_column = match column_iter.next() {
		Some(value) => value,
		None => return None
	};
	let timestamp = date_column.parse::<i64>();
	if timestamp.is_err() {
		// If the line cannot be parsed, and is the first line of file,
		// it is the header of CSV and we just return original to output
		if on_first_line {
			let new_line = column_iter.fold(date_column.to_string(), concat_columns);
			return Some(new_line);
		}
		// If not the first line, print error and continue to next line
		if !quiet {
			writeln!(&mut io::stderr(), "Line {} doesn't start with timestamp", line).unwrap();
		}
		return None;
	}
	let dest_datetime = dest_timezone.timestamp(timestamp.unwrap(), 0);
	// Filter against time_point
	if let Some(t) = time_point {
		if dest_datetime.time() != t { return None }
	}
	let new_line = column_iter.fold(dest_datetime.to_string(), concat_columns);
	Some(new_line)
}


fn main() -> Result<()> {
	let matches = App::new("Cleaning InfluxDB's export CSV")
		.version(crate_version!()).author(crate_authors!())
		.about("Clean CSV file, exported from InfluxDB.\n\
		       Remove the first column (measurement name) and convert timezone for time column")
		.arg(Arg::with_name("INPUT")
		     .help("Input file name. - for stdin.")
		     .required(true))
		.arg(Arg::with_name("quiet")
		     .short("q")
		     .help("Quiet. No error message when parsing CSV."))
		.arg(Arg::with_name("timezone")
		     .short("t")
		     .value_name("timezone name")
		     .help("Timezone (e.g. Asia/Ho_Chi_Minh) to convert to. \n
		           Original InfluxDB CSV has time in UTC."))
		.arg(Arg::with_name("time_point")
		     .short("p")
		     .value_name("time point")
		     .help("Filter and get only row at this time point (e.g. 07:00:00, of destination timezone) in a day."))
		.arg(Arg::with_name("output")
		     .short("o")
		     .value_name("file name")
		     .help("Output file name. Ommit to write to stdout."))
		.get_matches();

	color_eyre::install()?;
	let infile = matches.value_of("INPUT").unwrap();
	let dest_timezone = match matches.value_of("timezone") {
		Some(name) => name.parse::<Tz>().map_err(|e| eyre!(e))?,
		None => chrono_tz::UTC
	};
	let quiet = matches.is_present("quiet");

	let time_point = if matches.is_present("time_point") {
		let time_string = matches.value_of("time_point").unwrap();
		match NaiveTime::parse_from_str(time_string, "%H:%M:%S") {
			Ok(parsed_time) => Some(parsed_time),
			Err(_) => panic!("Invalid time string. Should be in form of HH:MM:SS.")
		}
	} else {None};

	let stdin = io::stdin();
	let reader = if infile == "-" {
		Box::new(stdin.lock()) as Box<dyn BufRead>
	} else {
		let f = File::open(infile).wrap_err_with(|| format!("Failed to open {} file!", infile))?;
		Box::new(BufReader::new(f))
	};

	let stdout = io::stdout();
	let mut writer: Box<dyn Write> = match matches.value_of("output") {
		Some(outfile) => {
			let created = File::create(&Path::new(outfile));
			assert!(created.is_ok(), "Failed. Cannot create file: {}", outfile);
			Box::new(BufWriter::new(created.unwrap()))
		},
		None => Box::new(stdout.lock())
	};

	for (i, wline) in reader.lines().enumerate() {
		let line = wline.unwrap();
		process_line(line, i == 0, dest_timezone, time_point, quiet)
			.map(|l| writeln!(&mut writer, "{}", l).unwrap());
	}
	Ok(())
}
